### PR TITLE
Add permission to secure access to certain config files

### PR DESCRIPTION
**Pull request by thecoop**
Add a security permission for code to be able to secure access to config files.

If a codebase grants itself `SecuredFileAccessPermission` on a file or directory, it prevents ES and other plugins from accessing that path at all. This can be used by a plugin to keep secrets stored in files safe from other code.

This removes the forbidden permissions added in https://github.com/elastic/elasticsearch/pull/107481, and replaces it with a secured access permission to nothing

### CODE REVIEW

**Comment by thecoop** [comment]
@rjernst @tvernum Requesting review on overall approach, will expand it out with proper tests if what is here is good in principle.

**Comment by thecoop** [comment]
I'm not sure how to go about testing this at the plugin level - all the existing plugin security tests just test the permissions that are read, they don't go all the way through to creating an `ESPolicy`/running it in ES

**Comment by rjernst** [comment]
> I'm not sure how to go about testing this at the plugin level

Could you add tests to ESPolicyTests which mimic plugins? That is, you can construct mock plugins (see eg how InMemoryModuleFinderTests uses JarUtils to create dynamic jars). They don't need to be "real" plugins in the sense they don't need to have a class inherit from Plugin, nor have a plugin descriptor. All we care about from the ESPolicy perspective is the map of codebases, and the mapping of those codebases to concrete Policy objects.

**Comment by thecoop** [comment]
I'm not sure that will test what I want - at the moment, there's nothing testing `Security` itself, in particular that the exclusive files are configured. Ideally I want something that tests if `Security` and `ESPolicy` together work to deny access to the relevant files

**Comment by rjernst** [comment]
> Ideally I want something that tests if Security and ESPolicy together work to deny access to the relevant files

That's a more difficult test since Security sets up the SM for the running VM. I'm suggesting adding a unit test (ESPolicyTests already exists) to directly check the behavior of `implies` given some set of arguments to the ctor.

**Comment by thecoop** [comment]
That's what I've already added to `ESPolicyUnitTests.testExclusiveAccess`. Sounds like this is only a manual test then.

**Comment by thecoop** [comment]
The permission is now all properly working, checked with a manual test

**Review by rjernst — COMMENTED** [review]
Test looks good. I was previously thinking we needed to pass in a map of plugins, but I see why it's not necessary now (the new check happens before looking at code location).

File: server/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java [hunk 1]
```diff
@@ -122,6 +155,28 @@ public boolean implies(ProtectionDomain domain, Permission permission) {
...
+
+        // check the source
+        Set<URL> accessibleSources = exclusiveFiles.get(permission);
+        if (accessibleSources != null) {
+            // simple case - single-file referenced directly
+            return accessibleSources.contains(location);
+        } else {
+            // there's a directory reference in there somewhere
+            // do a manual search :(
+            // there may be several permissions that potentially match
```
[the remark this answers is not in the document]
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/bootstrap/ESPolicy.java:172]: It is - a plugin (or ES) can gain access to an exclusive file simply by granting an ExclusiveAccessFilePermission to itself. A plugin can also grant several overlapping exclusive permissions (permission to a dir, and permission to specific files in that dir)

**Review by rjernst — APPROVED** [review]
LGTM

**Pull request merged**